File: c++/src/2000-2999/2654minOperations.cpp
#include "lc_pub.h"

class Solution {
    public:
    int minOperations(vector<int>& nums) {
        int g=nums[0], n=nums.size(),c1=0;
        for (auto x: nums) {
            g=gcd(g, x);
            if (x==1) c1++;
        }
        if (g > 1) return -1;
        if (c1) return n-c1;
        vector<pair<int, int>> g1;
        int mn=n;  // 最小能合并成1的子区间长度

        for (int i=0;i<n;i++) {
            int x=nums[i];
            for (int j=g1.size()-1;j>=0;j--) {
                g1[j].first=gcd(g1[j].first, x);
            }
            g1.push_back({x, i});
            // g1 去重
            int cur=0;
            for (int j=0;j<g1.size();j++) {
                if (g1[cur].first==g1[j].first) {
                    g1[cur]=g1[j];
                }
                else {
                    cur++;
                    g1[cur]=g1[j];
                }
                if (g1[cur].first==1) mn=min(mn, i-g1[cur].second+1);
            }
            g1.resize(cur+1);
        }
        return mn-1+n-1;
    }
    };

    
int main()
{
    cout<<"test let us start! %s" << __cplusplus <<std::endl;
    vector<int> arr = {2,6,3,4};

    Solution so;
    cout << so.minOperations(arr) << endl;
    return 0;
}

File: allcode/2600-2699/2654minOperations.py
# 给你一个下标从 0 开始的 正 整数数组 nums 。你可以对数组执行以下操作 任意 次：
#
# 选择一个满足 0 <= i < n - 1 的下标 i ，将 nums[i] 或者 nums[i+1] 两者之一替换成它们的最大公约数。
# 请你返回使数组 nums 中所有元素都等于 1 的 最少 操作次数。如果无法让数组全部变成 1 ，请你返回 -1 。
#
# 两个正整数的最大公约数指的是能整除这两个数的最大正整数。
#
#
#
# 示例 1：
#
# 输入：nums = [2,6,3,4]
# 输出：4
# 解释：我们可以执行以下操作：
# - 选择下标 i = 2 ，将 nums[2] 替换为 gcd(3,4) = 1 ，得到 nums = [2,6,1,4] 。
# - 选择下标 i = 1 ，将 nums[1] 替换为 gcd(6,1) = 1 ，得到 nums = [2,1,1,4] 。
# - 选择下标 i = 0 ，将 nums[0] 替换为 gcd(2,1) = 1 ，得到 nums = [1,1,1,4] 。
# - 选择下标 i = 2 ，将 nums[3] 替换为 gcd(1,4) = 1 ，得到 nums = [1,1,1,1] 。
# 示例 2：
#
# 输入：nums = [2,10,6,14]
# 输出：-1
# 解释：无法将所有元素都变成 1 。
#
#
# 提示：
#
# 2 <= nums.length <= 50
# 1 <= nums[i] <= 106
from math import gcd

from leetcode.allcode.competition.mypackage import *

class Solution:
    def minOperations(self, nums: List[int]) -> int:
        n = len(nums)
        if min(nums) == 1:
            return sum(1 for x in nums if x != 1)
        ans = inf
        for i in range(n):
            j = i
            cur = nums[i]
            while j < n:
                cur = gcd(cur, nums[j])
                if cur == 1:
                    break
                j += 1
            if cur == 1:
                ans = min(ans, j - i + n - 1)
        return ans if ans < inf else -1


so = Solution()
print(so.minOperations([1,1]))
print(so.minOperations([2,6,3,4]))   # 4
print(so.minOperations([2,10,6,14]))




